### PR TITLE
Use database for email verification temp storage

### LINKED ISSUE #59: Use database for keeping track of emails in the process of being verified
Will be needed when we scale.

### PR DESCRIPTION
**Pull request by nhnt11**
Fixes #59

### CODE REVIEW

**Comment by nhnt11** [comment]
Fixes #59

File: routes/user.js [hunk 1]
```diff
@@ -2,58 +2,72 @@
...
-        info: "sent verification link",
-        // Send the would-be link back to the client in dummy mode.
-        // eslint-disable-next-line no-process-env
-        link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
-      });
-    })
-    .catch(error => {
-      res.json({ email, info: "failed to send verification link", error });
-      throw error;
+  const url = `${AppConstants.SERVER_URL}/user/verify?state=${state}&email=${email}`;
```
Comment (pdehaan) [inline routes/user.js]: Sorry, i don't know the context of `state` or `email`, but do they need to be URI encoded?

File: routes/user.js [hunk 2]
```diff
@@ -2,58 +2,72 @@
...
+
+  try {
+    const { error } = await Subscribers.addTempUser(email, state);
+
+    if (error) {
+      res.json({ email, error: new Error("Database error.") });
```
Comment (pdehaan) [inline routes/user.js]: Not sure how "user facing" these errors may be, but are we sure that dumping an Error() object won't leak any useful information (like server file paths or stack traces)? Or should we just return a string, a la:

```js
res.json({ email, error: "Database error" });
```

File: routes/user.js [hunk 3]
```diff
@@ -2,58 +2,72 @@
...
+    res.json({
+      email,
+      info: "sent verification link",
+      // Send the would-be link back to the client in dummy mode.
+      // eslint-disable-next-line no-process-env
+      link: process.env.DEBUG_DUMMY_SMTP ? url : undefined,
     });
+  } catch (error) {
+    console.log(error);
+    res.json({ email, info: "failed to send verification link", error });
```
Comment (pdehaan) [inline routes/user.js]: Same as comment above, but should we be returning Error _objects_ to the user, or just the error messages?

```js
res.json({ email, info: "Failed to send verification link", error: error.message });
```

**Comment by pdehaan** [comment]
> Fixes #59

Unrelated to anything specific, but... this comment needs to be in your _original_ commit message (you can edit it above) if you want GitHub to auto-close the associated issue.

Sorry, I don't seem to have permissions in this repo, otherwise I could edit the comments directly.

File: routes/user.js (hunk 1 above)
Comment (nhnt11) [reply routes/user.js]: Good point thanks

File: routes/user.js (hunk 2 above)
Comment (nhnt11) [reply routes/user.js]: Yeah, I thought about this but promptly forgot. A string is probably better here. In any case this should be replaced by an error code (#65)

File: subscribers.js [hunk 4]
```diff
@@ -73,6 +74,74 @@ const Subscribers = {
...
+      return _ret(e);
+    }
+  },
+
+  async getTempUser(aEmail) {
+    try {
+      const rows = (await dbq({
+        text: "SELECT * FROM users_temp WHERE email = $1;",
+        values: [ aEmail ],
+      })).rows;
```
Comment (pdehaan) [inline subscribers.js]: Per https://github.com/mozilla/blurts-server/pull/70#discussion_r173312299

> Not sure if we could just destructure `const {rows}` here, and save some brackets and `.rows` property. Maybe.
>
> ```js
> const {rows} = await dbq({
>   text: "SELECT * FROM users_temp WHERE email = $1;",
>   values: [ aEmail ],
> });
> ```

File: routes/user.js [hunk 5]
```diff
@@ -2,58 +2,72 @@
...
+    if (token !== req.query.state) {
+      res.json({ error: new Error("Token and email don't match."), email });
+      return;
+    }
+
+    await Subscribers.addUser(email);
+    res.json({ email, info: `Successfully added ${email}` });
+  } catch (error) {
+    console.log(error);
+    res.json({ info: "Failed to add email to database.", error });
```
Comment (pdehaan) [inline routes/user.js]: Should this also include the `email` property?

File: db-utils.js [hunk 6]
```diff
@@ -1,21 +1,45 @@
...
+    // Use PG* env vars to configure client.
+    const client = new pg.Client();
+    try {
+      await client.connect();
+      await client.query("DROP TABLE IF EXISTS users_temp;");
+      await client.query("CREATE TABLE users_temp ( \
+                            email VARCHAR(320) UNIQUE, \
+                            token VARCHAR(80), \
+                            time_added TIMESTAMP DEFAULT NOW() \
+                          );");
```
Comment (pdehaan) [inline db-utils.js]: I think if you use template strings with a backtick (``` ` ```), you can avoid doing the multiline backslash trick.

```js
      await client.query(`CREATE TABLE users_temp (
                            email VARCHAR(320) UNIQUE,
                            token VARCHAR(80),
                            time_added TIMESTAMP DEFAULT NOW()
                          );`);
```

Comment (nhnt11) [reply db-utils.js]: Fixed in #70, let's continue the conversation there?

File: routes/user.js (hunk 5 above)
Comment (nhnt11) [reply routes/user.js]: Getting rid of returning the email in the response in #70.

File: subscribers.js (hunk 4 above)
Comment (nhnt11) [reply subscribers.js]: Fixed this in a commit on #70.

File: routes/user.js (hunk 3 above)
Comment (nhnt11) [reply routes/user.js]: #70 fixes this.

**Comment by nhnt11** [comment]
Let's close this and just do it in #70

**Pull request closed without merging**